### DISCUSSION TITLE
Referencing EObjects which haven't got a name - the right way

### DISCUSSION THREAD

**Discussion by tfonda-fbk** [eclipse-xtext/xtext#3181] [Q&A] [answered]
Hi,

In the DSL I'm working on I want to allow entities to override other entities. See the example grammar and entity declarations presented here: https://github.com/eclipse/xtext-website/pull/55

The main point is that I want to be able to cross-reference some entities which have got no name. These entities do however override some other entities (see the grammar: an entity may not have a name only if it overrides another entity), so my first naive idea was to implement a custom IDerivedStateComputer that sets the `name` field to the name of the overridden entity, for all entities without a name. This seemed to work fine, but @szarnekow suggested that this solution is fragile and poorly designed, so I had to come up with a different and more solid solution to my problem.

I have now solved it in a completely different way that does not require resolving any cross-reference and/or tampering with the `name` feature of EObjects. My new solution makes use of a custom IQualifiedNameProvider to build a qualified name, falling back on using the name of the overridden entity when an entity hasn’t got its own name. I used the `NodeModelUtils.findNodesForFeature()` and then the `NodeModelUtils.getTokenText()` methods to extract the overridden entity’s name from the parsed DSL file without having to solve any cross-reference. Finally, I’ve overridden `computeFullyQualifiedNameFromNameAttribute()` to achieve the desired, aforementioned behavior.

I then had to customize the scope provider to use the last segment of an EObject’s qualified name as the identifier of that EObject for solving cross-references. This way, nothing changes for all EObjects with a name, while additionally I can finally reference EObjects of type Entity which override another Entity but have got no name. An excerpt of my code follows:
```java
return Scopes.scopeFor(iterable,
	QualifiedName.wrapper(eObject -> {
		final QualifiedName qualifiedName = qualifiedNameProvider.getFullyQualifiedName(eObject);
		if (qualifiedName == null) return null;
		return qualifiedName.getLastSegment();
	}),
	IScope.NULLSCOPE);
```

My question is: does this look like a proper way to achieve what I want to achieve? Any feedback on it would be very welcome :)
Thanks!

**Comment by cdietrich** [comment]
Can you elaborate your scope provider logic.
Do you refer to them always manually collected or also globally 

besides that it makes sense
Besides computeFullyQualifiedNameFromNameAttribute

But this depends on what you exactly do

**Reply by tfonda-fbk** [reply]
@cdietrich 
You can find my full scope and qualified name providers here: https://gist.github.com/tfonda-fbk/5fb6d8cfef99fefc9d17020b037008ab

The DSL I'm implementing is (a fragment of) the SysML v2 textual notation.
To clarify something about my qualified name provider: in my metamodel, Usage is the only kind of entity allowed not to have a name but to redefine (override) another Usage. The redefinition is contained in the RedefinitionShorthand AST object.
In my project I'm also using namespace imports (global scoping), if this is what you were asking.
Please let me know if any part of my code is not clear.
Thanks!

**Reply by cdietrich** [reply, marked as answer]
i guess then the solution is fine
